{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import infercnvpy as cnv\n",
    "import matplotlib.pyplot as plt\n",
    "# make this notebook work better with Scanpy\n",
    "import warnings; warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyensembl import EnsemblRelease\n",
    "\n",
    "# release 98 uses human reference genome GRCh38\n",
    "data = EnsemblRelease(98)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make output directories\n",
    "import os\n",
    "if not os.path.exists(\"ST_out\"):\n",
    "    os.mkdir(\"ST_out\")\n",
    "\n",
    "if not os.path.exists(\"ST_out/infercnv/\"):\n",
    "    os.mkdir(\"ST_out/infercnv/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(transparent=True, dpi_save=180)\n",
    "sc.settings.figdir = \"ST_out/infercnv/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "### Read in key dataframe with sample information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_key = pd.read_csv(\"../resources/ST/visium_sample_key.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Perform CNV inference by patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_key.CNV_group.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pat in sample_key.CNV_group.unique():\n",
    "    try:\n",
    "        print(\"Starting on patient {}\".format(pat))\n",
    "        key_tmp = sample_key.loc[sample_key.CNV_group == pat, :].copy()\n",
    "        outs = []\n",
    "        for s in key_tmp.index:\n",
    "            a = sc.read(\"../data/ST/{}_master.h5ad\".format(s))\n",
    "            print(\"Read adata from ../data/ST/{}_master.h5ad\".format(s))\n",
    "            outs.append(a)\n",
    "        # concatenate anndata objects\n",
    "        a_comb = outs[0].concatenate(\n",
    "            outs[1:],\n",
    "            join=\"outer\",\n",
    "            batch_categories=list(key_tmp.index),\n",
    "            fill_value=0,\n",
    "        )\n",
    "        del a_comb.obsm\n",
    "        \n",
    "        # set up for InferCNV\n",
    "        a_comb.var.drop(columns=a_comb.var.columns[a_comb.var.columns.str.startswith(\"gene_ids\")][1:], inplace=True)\n",
    "        a_comb.var.rename(columns={a_comb.var.columns[a_comb.var.columns.str.startswith(\"gene_ids\")][0]:\"gene_ids\"}, inplace=True)\n",
    "        # Use `pyensembl` to map ENSG IDs to genomic loci\n",
    "        a_comb.var[\"chromosome\"] = \"0\"\n",
    "        a_comb.var[\"start\"] = 0\n",
    "        a_comb.var[\"end\"] = 0\n",
    "        err_counter = 0\n",
    "        for i in a.var_names:\n",
    "            try:\n",
    "                gene = data.gene_by_id(a_comb.var.gene_ids[i])\n",
    "                a_comb.var.loc[i,\"chromosome\"] = \"chr{}\".format(gene.contig)\n",
    "                a_comb.var.loc[i,\"start\"] = gene.start\n",
    "                a_comb.var.loc[i,\"end\"] = gene.end\n",
    "            except ValueError as e:\n",
    "                #print(e)\n",
    "                err_counter += 1\n",
    "        print(\"{} errors in ENSEMBL mapping\".format(err_counter))\n",
    "        \n",
    "        # run InferCNV\n",
    "        # provide stromal regions as \"normal cells\"\n",
    "        cnv.tl.infercnv(\n",
    "            a_comb,\n",
    "            reference_key=\"pathology_annotation\",\n",
    "            reference_cat=list(set([\"normal_mucosa\",\"smooth_muscle\"]).intersection(set(a_comb.obs.pathology_annotation.unique()))),\n",
    "            window_size=200,\n",
    "        )\n",
    "        # determine clusters based on detected CNVs\n",
    "        cnv.tl.pca(a_comb)\n",
    "        cnv.pp.neighbors(a_comb, n_neighbors=int(np.sqrt(a_comb.n_obs)))\n",
    "        cnv.tl.leiden(a_comb, resolution=1.0)\n",
    "        # score detected CNVs in every spot\n",
    "        cnv.tl.cnv_score(a_comb)\n",
    "        # get rid of noisy normal clusters\n",
    "        m = {\n",
    "            \"B\" :\n",
    "            list(a_comb.obs.loc[a_comb.obs.cnv_score <= a_comb.obs.cnv_score.quantile([.33]).values[0], \"cnv_leiden\"].unique())\n",
    "        }\n",
    "        m2 = {v: k for k,vv in m.items() for v in vv}\n",
    "        a_comb.obs.cnv_leiden = a_comb.obs.cnv_leiden.replace(m2)\n",
    "        \n",
    "        # plotting\n",
    "        print(\"Plotting pathology_annotation CNV heatmap\")\n",
    "        \n",
    "        a_comb.obs[\"pathology_annotation\"] = a_comb.obs[\"pathology_annotation\"].astype(str)\n",
    "        a_comb.obs[\"pathology_annotation\"] = a_comb.obs[\"pathology_annotation\"].fillna(\"smooth_muscle\")\n",
    "        a_comb.obs[\"pathology_annotation\"] = a_comb.obs[\"pathology_annotation\"].astype(\"category\")\n",
    "        # rename pathology_annotation categories and get proper colors\n",
    "        path_dict = {\n",
    "            \"carcinoma_edge\":\"carcinoma_border\",\n",
    "            \"UNK\":\"carcinoma\",\n",
    "            \"necrotic\":\"carcinoma\",\n",
    "            \"adipose\":\"smooth_muscle\",\n",
    "            \"nan\":\"smooth_muscle\",\n",
    "        }\n",
    "        a_comb.obs[\"pathology_annotation\"].replace(path_dict, inplace=True)\n",
    "        path_colordict = dict(zip(\n",
    "            [\"adenoma\",\"adenoma_border\",\"carcinoma\",\"carcinoma_border\",\"lymphoid_follicle\",\"normal_mucosa\",\"smooth_muscle\"],\n",
    "            sns.color_palette(\"tab10\",7).as_hex()\n",
    "        ))\n",
    "        a_comb.uns[\"pathology_annotation_colors\"] = [path_colordict[x] for x in a_comb.obs[\"pathology_annotation\"].cat.categories]\n",
    "        \n",
    "        # plot heatmap with tissue domains\n",
    "        cnv.pl.chromosome_heatmap(\n",
    "            a_comb,\n",
    "            groupby=\"pathology_annotation\",\n",
    "            save=\"_pathology_annotation_{}.png\".format(pat),\n",
    "            dendrogram=True,\n",
    "            figsize=(12,8),\n",
    "        )\n",
    "        if len(key_tmp.patient_name.unique()) == 1:\n",
    "            if key_tmp.patient_name.unique()[0] == \"HTA11_08622\":\n",
    "                # plot heatmap with WES ROIs\n",
    "                print(\"Plotting CNV heatmap by block (HTA11_08622 synchronous polyps)\")\n",
    "                cnv.pl.chromosome_heatmap(\n",
    "                    a_comb,\n",
    "                    groupby=\"Block ID\",\n",
    "                    save=\"_block_{}.png\".format(pat),\n",
    "                    dendrogram=True,\n",
    "                    figsize=(12,8),\n",
    "                )\n",
    "            else:\n",
    "                if \"LCM_ROI\" in a_comb.obs.columns:\n",
    "                    # plot heatmap with WES ROIs\n",
    "                    print(\"Plotting LCM_ROI CNV heatmap\")\n",
    "                    a_comb.obs.LCM_ROI = a_comb.obs.LCM_ROI.astype(str)\n",
    "                    cnv.pl.chromosome_heatmap(\n",
    "                        a_comb[a_comb.obs.LCM_ROI != \"nan\", :],\n",
    "                        groupby=\"LCM_ROI\",\n",
    "                        save=\"_LCM_ROI_{}.png\".format(pat),\n",
    "                        dendrogram=True,\n",
    "                        figsize=(12,8),\n",
    "                    )\n",
    "            # plot heatmap with CNV Leiden clusters\n",
    "            print(\"Plotting CNV heatmap with Leiden clusters\")\n",
    "            cnv.pl.chromosome_heatmap(\n",
    "                a_comb,\n",
    "                groupby=\"cnv_leiden\",\n",
    "                save=\"_{}.png\".format(pat),\n",
    "                dendrogram=True,\n",
    "                figsize=(12,8),\n",
    "            )\n",
    "        elif len(key_tmp.patient_name.unique()) > 1:\n",
    "            # plot heatmap with WES ROIs\n",
    "            print(\"Plotting CNV heatmap by patient (for grouped samples)\")\n",
    "            cnv.pl.chromosome_heatmap(\n",
    "                a_comb,\n",
    "                groupby=\"Patient\",\n",
    "                save=\"_patient_{}.png\".format(pat),\n",
    "                dendrogram=True,\n",
    "                figsize=(12,8),\n",
    "            )\n",
    "            for pat2 in key_tmp.patient_name.unique():\n",
    "                # plot heatmap with CNV Leiden clusters\n",
    "                print(\"Plotting CNV heatmap with Leiden clusters for {}\".format(pat2))\n",
    "                cnv.pl.chromosome_heatmap(\n",
    "                    a_comb[a_comb.obs.Patient == pat2, :],\n",
    "                    groupby=\"cnv_leiden\",\n",
    "                    save=\"_{}_{}.png\".format(pat, pat2),\n",
    "                    dendrogram=True,\n",
    "                    figsize=(12,8),\n",
    "                )\n",
    "                \n",
    "        # rename MILWRM domain and ensure proper coloring\n",
    "        #a_comb.obs.rename(columns={\"VUMCrefNMF30_MILWRM_domain\":\"MILWRM Domain\"}, inplace=True)\n",
    "        a_comb.obs[\"MILWRM Domain\"] = a_comb.obs[\"MILWRM Domain\"].astype(\"category\")\n",
    "        mw_colordict = dict(zip([\"D0\",\"D1\",\"D2\",\"D3\",\"D4\",\"D5\",\"D6\",\"D7\"], sns.color_palette(\"plasma\",8).as_hex()))\n",
    "        a_comb.uns[\"MILWRM Domain_colors\"] = [mw_colordict[x] for x in a_comb.obs[\"MILWRM Domain\"].cat.categories]\n",
    "        \n",
    "        # plot PCA of patient/group\n",
    "        print(\"Plotting CNV PCA\")\n",
    "        sc.pl.embedding(\n",
    "            a_comb,\n",
    "            basis=\"X_cnv_pca\",\n",
    "            color=[\"pathology_annotation\", \"MILWRM Domain\", \"cnv_leiden\"],\n",
    "            save=\"_{}.png\".format(pat),\n",
    "            ncols=1,\n",
    "            frameon=False,\n",
    "        )\n",
    "        \n",
    "        a_comb.obs.cnv_leiden = a_comb.obs.cnv_leiden.astype(\"category\")\n",
    "        # save CNV cluster colors\n",
    "        cnv_leiden_colordict = dict(zip(list(a_comb.obs.cnv_leiden.cat.categories), a_comb.uns[\"cnv_leiden_colors\"]))\n",
    "        np.save(\"ST_out/infercnv/cnv_leiden_colors_{}.npy\".format(pat), cnv_leiden_colordict)\n",
    "        \n",
    "        # save CNV genome partitioning for patient to file\n",
    "        np.save(\"ST_out/infercnv/uns_cnv_{}.npy\".format(pat), a_comb.uns[\"cnv\"])\n",
    "        \n",
    "        # plot on Visium scaffolds\n",
    "        print(\"Plotting spatial overlays\")\n",
    "        for i, out in enumerate(outs):\n",
    "            # write to csv for compilation\n",
    "            tmp_obs = a_comb.obs.loc[a_comb.obs.batch == list(key_tmp.index)[i], [\"cnv_leiden\",\"cnv_score\"]].copy()\n",
    "            tmp_obs.index = out.obs_names\n",
    "            tmp_obs.to_csv(\"ST_out/infercnv/{}_cnv_leiden.csv\".format(list(key_tmp.index)[i]))\n",
    "            \n",
    "            # write CNV values to npz for compilation\n",
    "            np.savez_compressed(\"ST_out/infercnv/{}_cnv.npz\".format(list(key_tmp.index)[i]), a_comb[a_comb.obs.batch == list(key_tmp.index)[i], :].obsm[\"X_cnv\"])\n",
    "            \n",
    "            # add cnv metadata to individual sample anndata\n",
    "            out.obs[[\"cnv_leiden\",\"cnv_score\"]] = a_comb.obs.loc[a_comb.obs.batch == list(key_tmp.index)[i], [\"cnv_leiden\",\"cnv_score\"]].values\n",
    "            out.obs.cnv_score = out.obs.cnv_score.astype(float)\n",
    "            out.obs.cnv_leiden = out.obs.cnv_leiden.astype(\"category\")\n",
    "            \n",
    "            # rename pathology_annotation categories and get proper colors\n",
    "            out.obs[\"pathology_annotation\"].replace(path_dict, inplace=True)\n",
    "            out.uns[\"pathology_annotation_colors\"] = [path_colordict[x] for x in out.obs[\"pathology_annotation\"].cat.categories]\n",
    "            \n",
    "            # transfer cnv_leiden colors to out\n",
    "            out.uns[\"cnv_leiden_colors\"] = [cnv_leiden_colordict[x] for x in out.obs.cnv_leiden.cat.categories]\n",
    "            \n",
    "            # plot spatial\n",
    "            sc.pl.spatial(\n",
    "                out,\n",
    "                color=[\"cnv_score\",\"cnv_leiden\",\"pathology_annotation\"],\n",
    "                size=1.7,\n",
    "                ncols=3,\n",
    "                img_key=None,\n",
    "                frameon=False,\n",
    "                vmin=0.0,\n",
    "                vmax=a_comb.obs.cnv_score.max(),\n",
    "                cmap=\"viridis\",\n",
    "                save=\"_{}_{}_CNV.png\".format(pat, list(key_tmp.index)[i]),\n",
    "            )\n",
    "    except ValueError as e:\n",
    "        print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Get `infercnv` outputs to add to master AnnDatas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in sample_key.loc[sample_key.patient_name.isin(sample_key.patient_name.unique()), :].index:\n",
    "    print(\"Starting {}:\".format(s), end=\"\\n\\t\")\n",
    "    a = sc.read(\"../data/ST/{}_master.h5ad\".format(s))\n",
    "    \n",
    "    # drop old CNV column(s)\n",
    "    a.obs.drop(columns=a.obs.columns[a.obs.columns.str.startswith(\"cnv_leiden\")], inplace=True)\n",
    "    a.obs.drop(columns=a.obs.columns[a.obs.columns.str.startswith(\"cnv_score\")], inplace=True)\n",
    "    if \"CNV Clone\" in a.obs.columns:\n",
    "        a.obs.drop(columns=[\"CNV Clone\"], inplace=True)\n",
    "    if \"CNV Score\" in a.obs.columns:\n",
    "        a.obs.drop(columns=[\"CNV Score\"], inplace=True)\n",
    "        \n",
    "    # get InferCNV outputs\n",
    "    cnv = pd.read_csv(\"ST_out/infercnv/{}_cnv_leiden.csv\".format(s), index_col=0)\n",
    "    print(\"Read InferCNV outputs from ST_out/infercnv/{}_cnv_leiden.csv\".format(s))\n",
    "    a.obs = a.obs.merge(cnv, left_index=True, right_index=True)\n",
    "    \n",
    "    # loading uns dict back in from .npy file\n",
    "    a.obs.cnv_leiden = a.obs.cnv_leiden.astype(str)\n",
    "    a.obs.cnv_leiden = a.obs.cnv_leiden.astype(\"category\")\n",
    "    cnv_leiden_colordict = np.load(\"ST_out/infercnv/cnv_leiden_colors_{}.npy\".format(sample_key.loc[s, \"CNV_group\"]), allow_pickle='TRUE').item()\n",
    "    a.uns[\"cnv_leiden_colors\"] = [cnv_leiden_colordict[x] for x in a.obs.cnv_leiden.cat.categories]\n",
    "    \n",
    "    # save to master anndata object\n",
    "    print(\"\\tSaving to ../data/ST/{}_master.h5ad\".format(s), end=\"\\n\\n\")\n",
    "    a.write(\"../data/ST/{}_master.h5ad\".format(s), compression=\"gzip\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
